dashboard/app/utils/fine_tuning/fireworks.ts

```diff
@@ -0,0 +1,558 @@
+ * result handling, and deployment management.
+ */
+
+import type { SFTFormValues } from "~/routes/optimization/fine-tuning/types";
+import type { JsExposedEnv } from "../minijinja/pkg/minijinja_bindings";
+import { v7 } from "uuid";
+import type {
+  ContentBlockOutput,
+  JsonInferenceOutput,
+  ParsedInferenceRow,
+} from "../clickhouse";
+import { getCuratedInferences } from "../clickhouse";
+import { render_message } from "./rendering";
+import { getConfig } from "../config/index.server";
+import { get_template_env, type ChatCompletionConfig } from "../config/variant";
+import { z } from "zod";
+import { SFTJob, type SFTJobStatus } from "./common";
+import type { ProviderType } from "../config/models";
+export const FIREWORKS_API_URL = "https://api.fireworks.ai";
+export const FIREWORKS_API_KEY = process.env.FIREWORKS_API_KEY || throwError();
+export const FIREWORKS_ACCOUNT_ID =
+  process.env.FIREWORKS_ACCOUNT_ID || throwError();
+
+// This is apparently the traditional way to coerce both to strings.
+function throwError(): never {
+  throw new Error("FIREWORKS_API_KEY and FIREWORKS_ACCOUNT_ID must be set");
+}
+
+interface FireworksSFTJobParams {
+  jobPath: string;
+  status: string;
+  jobId: string;
+  modelId?: string;
+  modelPath?: string;
+}
+
+export class FireworksSFTJob extends SFTJob {
+  public jobPath: string;
+  public jobStatus: string;
+  public jobId: string;
+  public modelId?: string;
+  public modelPath?: string;
+
+  constructor(params: FireworksSFTJobParams) {
+    super();
+    this.jobPath = params.jobPath;
+    this.jobStatus = params.status;
+    this.jobId = params.jobId;
+    this.modelId = params.modelId;
+    this.modelPath = params.modelPath;
+  }
+
+  static async from_form_data(data: SFTFormValues): Promise<FireworksSFTJob> {
+    const config = await getConfig();
+    // TODO: throw if this isn't a chat completion
+    const currentVariant = config.functions[data.function].variants[
+      data.variant
+    ] as ChatCompletionConfig;
+    const curatedInferences = await getCuratedInferences(
+      data.function,
+      config.functions[data.function],
+      data.metric,
+      config.metrics[data.metric],
+      data.maxSamples,
+    );
+    if (!curatedInferences || curatedInferences.length === 0) {
+      throw new Error("No curated inferences found");
+    }
+    const templateEnv = await get_template_env(currentVariant);
+    const jobPath = await start_sft_fireworks(
+      data.model.name,
+      curatedInferences,
+      data.validationSplitPercent,
+      templateEnv,
+    );
+    return new FireworksSFTJob({
+      jobPath: jobPath,
+      status: "RUNNING",
+      jobId: data.jobId,
+      modelId: undefined,
+      modelPath: undefined,
+    });
+  }
+
+  provider(): ProviderType {
+    return "fireworks";
+  }
+
+  status(): SFTJobStatus {
+    if (this.jobStatus === "FAILED") return "error";
+    return this.jobStatus === "COMPLETED" ? "completed" : "running";
+  }
```

Comment:
![CleanShot 2024-12-21 at 18 05 36@2x](https://github.com/user-attachments/assets/32bb7ac3-5fce-4411-b6be-cbe7e6092ef7)

After completing the job, we're stuck on running, likely because we get another status?

Reply:
Don't default to running, default to "unknown" or something. Log a warning here.

Reply:
fixed this for now, in the progress PR I'll do a thorough solution